src/emit/expression.rs

```diff
@@ -1904,106 +1904,178 @@ fn runtime_cast<'a>(
     val: BasicValueEnum<'a>,
     ns: &Namespace,
 ) -> BasicValueEnum<'a> {
-    if matches!(from, Type::Address(_) | Type::Contract(_))
-        && matches!(to, Type::Address(_) | Type::Contract(_))
-    {
-        // no conversion needed
-        val
-    } else if let Type::Address(_) = to {
-        let llvm_ty = bin.llvm_type(from, ns);
+    match (from, to) {
+        (Type::Bytes(_) | Type::Int(_) | Type::Uint(_) | Type::Value, Type::Address(_)) => {
```

Comment:
I think there is no more `Type::Value` after codegen.

Reply:
We need the higher-level type information after codegen, in order to generate the correct debug information. Also, in program runtime v2 the debug information becomes an essential part of the interface (via BTF). We can't drop higher-level type information.